EZP-27713: As a Developer I want HttpCache to handle RemoveTranslationSignal

alongosz opened the pull request:
Implements [EZP-27713](https://jira.ez.no/browse/EZP-27713)
----

This PR implements HttpCache `RemoveTranslationSlot` for `RemoveTranslationSignal` introduced in ezsystems/ezpublish-kernel#2031.

Similar slot for deprecated EzPublishCoreBundle HttpCache was introduced in ezsystems/ezpublish-kernel#2062.

**TODO**:
- [x] Implement `RemoveTranslationSlot` for HttpCache.
- [x] Add unit test for new Slot.
- [x] Perform negative (no Slot) test to confirm this solution works in real env as expected.
- [x] Get feedback on `composer.json` changes

alongosz commented on composer.json:
```diff
@@ -11,7 +11,7 @@
     "minimum-stability": "stable",
     "require": {
-        "ezsystems/ezpublish-kernel": "^6.7",
+        "ezsystems/ezpublish-kernel": "^6.7@dev",
```
I'm not sure if this is the right place or the right way to require `dev-master` for `ezpublish-kernel` (took that from `ezplatform` repository), but I need this as translation removal is a part of unreleased `1.11` / `6.11`.

andrerom replied:
This is ok, we do the same in solr: https://github.com/ezsystems/ezplatform-solr-search-engine/blob/master/composer.json#L15

alongosz commented on tests/SignalSlot/RemoveTranslationSlotTest.php:
```diff
@@ -0,0 +1,111 @@
+    public function setUp()
+    {
+        $this->persistenceHandlerMock = $this->getMock(PersistenceHandler::class);
+        $this->purgeClientMock = $this->getMock(PurgeClientInterface::class);
```
Parent `setUp` mocks deprecated interface, so I've overridden the whole method.

andrerom replied:
why not just fix parent? Assuming it is the purger the code uses.

alongosz replied:
I expected all classes inheriting from parent to require fixes, but this apparently is not the case, so fixed in 0075477.

alongosz commented on src/SignalSlot/RemoveTranslationSlot.php:
```diff
@@ -0,0 +1,76 @@
+     */
+    public function __construct(
+        PurgeClientInterface $purgeClient,
+        PersistenceHandler $persistenceHandler
```
`RemoveTranslationSignal` does not carry information about locations, so I had do make dependency on `PersistenceHandler`. Not sure if this is the optimal way, though. Any POV?
Just in case pushed whole location fetching as a fixup: 21caa86 so can be easily removed.

andrerom replied:
why do you need to clear location id? you can just clear by content id, and it will clear all relevant cache items. This is one of the many points with multi tagging, you don't need to always refer to cache by location id.

andrerom replied:
But ok for parent I see the point, but unsure if it is needed.

alongosz replied:
Answered [here](#r131672666) - comment on relevant part of the code

alongosz commented:
Tested manually with eZ Platform.
Review ping @andrerom @Nattfarinn (can't request a review for this repo :) )

alongosz commented on src/SignalSlot/RemoveTranslationSlot.php:
```diff
@@ -0,0 +1,76 @@
+            $signal->contentId
+        );
+        foreach ($locations as $location) {
+            $tags[] = 'location-' . $location->id;
```
> why do you need to clear location id? you can just clear by content id, and it will clear all relevant cache items. This is one of the many points with multi tagging, you don't need to always refer to cache by location id.

Interesting thing, the moment I comment-out this line, it stopped working returning cached version (checked using [this test](https://github.com/ezsystems/ezpublish-kernel/pull/2065/commits/011206b414f2cbae600533f069701da841159c54)). Don't really know anything about tagging to tell you why :(

> But ok for parent/siblings I see the point, but unsure if it is needed.

However I can't think of - looking at [REST API v2 spec](https://github.com/ezsystems/ezpublish-kernel/blob/master/doc/specifications/rest/REST-API-V2.rst) - any case when we can access Version via parent location. That would be a good case to check if `parent-` tag is needed there.

alongosz commented:
I just noticed I can use directly `LocationHandler` instead of `PersistenceHandler` as it is done in `PublishVersionSlot`.

In 9db1fe9 I updated `RemoveTranslationSlot` which optimizes performance. I also removed 
`'parent-'` tag as I can't think of any case it's going to be relevant for this Slot.

However without `'location-'` Slot won't work properly. Judging from other Content and Version changing Slots like `DeleteContentSlot` or `PublishVersionSlot` location is needed there.

Mentioned changes allowed me to simplify mocks, updated in a2de543.

andrerom commented:
> I just noticed I can use directly LocationHandler instead of PersistenceHandler as it is done in PublishVersionSlot.

Ok, if it is needed there it will be needed here also, so keep it aligned with that given this effectively is a re-publishing of the content.

alongosz commented:
> Ok, if it is needed there it will be needed here also, so keep it aligned with that given this effectively is a re-publishing of the content.

Perfectly aligned with `PublishVersionSlot` in 7b201b3 fixup, so there's no doubt about reasons ;) 
Also added inline comment to indicate that.
Tested with eZ Platform on `master`, `1.10`, `1.11`

Let me know when I can cleanup commits after review.

Review reping @Nattfarinn :P

alongosz commented:
Squashed fixups after reviews.

andrerom commented on tests/SignalSlot/RemoveTranslationSlotTest.php:
```diff
@@ -0,0 +1,96 @@
+
+    public function createSignal()
+    {
+        return new RemoveTranslationSignal(
```
Oh, shouldn't we skip the test if class is not present @alongosz? otherwise it will fail on prev branches right?

alongosz replied:
Ok, Added 83c47b0 fixup. For more details please see [this comment](#discussion_r132164390).

alongosz commented on tests/SignalSlot/RemoveTranslationSlotTest.php:
```diff
@@ -0,0 +1,106 @@
+    {
+        if (!class_exists(RemoveTranslationSignal::class)) {
+            self::markTestSkipped('RemoveTranslationSignal does not exist');
+        }
```
> Oh, shouldn't we skip the test if class is not present @alongosz? otherwise it will fail on prev branches right?

Since this is unit test run directly from this package, composer install will always install newest kernel, so this shouldn't really happen. But just to be safe and consistent (we require ezpublish-kernel `^6.7`) I added this skip in 83c47b0.

Notes:
- done in `setUpBeforeClass` because data providers also rely on some usages of the Signal.
- I've checked on 6.7.x that if this class does not exist `class_exists(RemoveTranslationSignal::class)` returns false without warnings. This is ["documented" behavior](https://www.reddit.com/r/lolphp/comments/6fwu28/classnameclass_is_valid_for_any_identifier/) :laughing: 
Buf if you'd feel safer with string FQCN, I can fix that :)

ping @andrerom @Nattfarinn

andrerom replied:
> Buf if you'd feel safer with string FQCN, I can fix that :)

no, it's fine like this, it's how it is supposed to work actually. Use is not an import :)
